File: markback/linter.py
"""MarkBack V2 linter implementation."""

import json
from pathlib import Path
from typing import Optional

from .parser import parse_file, parse_string
from .types import (
    Diagnostic,
    ErrorCode,
    ParseResult,
    Record,
    Severity,
    WarningCode,
    parse_feedback,
)
from .writer import _write_record_canonical, write_string


def lint_feedback_json(
    feedback: str,
    file: Optional[Path],
    line: Optional[int],
    record_idx: Optional[int],
) -> list[Diagnostic]:
    """Lint JSON-formatted feedback."""
    diagnostics: list[Diagnostic] = []

    if feedback.startswith("json:"):
        json_str = feedback[5:]
        try:
            json.loads(json_str)
        except json.JSONDecodeError as e:
            diagnostics.append(Diagnostic(
                file=file,
                line=line,
                column=None,
                severity=Severity.ERROR,
                code=ErrorCode.E007,
                message=f"Invalid JSON after json: prefix: {e}",
                record_index=record_idx,
            ))

    return diagnostics


def lint_feedback_structured(
    feedback: str,
    file: Optional[Path],
    line: Optional[int],
    record_idx: Optional[int],
) -> list[Diagnostic]:
    """Lint structured feedback for unclosed quotes."""
    diagnostics: list[Diagnostic] = []

    in_quote = False
    escaped = False

    for char in feedback:
        if escaped:
            escaped = False
            continue
        if char == '\\':
            escaped = True
            continue
        if char == '"':
            in_quote = not in_quote

    if in_quote:
        diagnostics.append(Diagnostic(
            file=file,
            line=line,
            column=None,
            severity=Severity.ERROR,
            code=ErrorCode.E008,
            message="Unclosed quote in structured attribute value",
            record_index=record_idx,
        ))

    return diagnostics


def lint_file_exists(
    record: Record,
    base_path: Optional[Path],
    record_idx: int,
) -> list[Diagnostic]:
    """Check if @file reference exists."""
    diagnostics: list[Diagnostic] = []

    if record.file and not record.file.is_uri:
        try:
            resolved = record.file.resolve(base_path)
            if not resolved.exists():
                diagnostics.append(Diagnostic(
                    file=record._source_file,
                    line=record._start_line,
                    column=None,
                    severity=Severity.WARNING,
                    code=WarningCode.W003,
                    message=f"@file not found: {record.file}",
                    record_index=record_idx,
                ))
        except ValueError:
            pass

    return diagnostics


def lint_input_exists(
    record: Record,
    base_path: Optional[Path],
    record_idx: int,
) -> list[Diagnostic]:
    """Check if @input reference exists."""
    diagnostics: list[Diagnostic] = []

    if record.input and not record.input.is_uri:
        try:
            resolved = record.input.resolve(base_path)
            if not resolved.exists():
                diagnostics.append(Diagnostic(
                    file=record._source_file,
                    line=record._start_line,
                    column=None,
                    severity=Severity.WARNING,
                    code=WarningCode.W009,
                    message=f"@input not found: {record.input}",
                    record_index=record_idx,
                ))
        except ValueError:
            pass

    return diagnostics


# V1 compat aliases
lint_source_exists = lint_file_exists
lint_prior_exists = lint_input_exists


def _is_position_invalid(ref) -> tuple[bool, str]:
    """Check if a FileRef has an invalid position range."""
    if ref.start_line is None or ref.end_line is None:
        return False, ""

    if ref.end_line < ref.start_line:
        return True, f"end line {ref.end_line} is less than start line {ref.start_line}"

    if ref.end_line == ref.start_line:
        if (ref.start_column is not None and
            ref.end_column is not None and
            ref.end_column < ref.start_column):
            return True, f"end column {ref.end_column} is less than start column {ref.start_column} on line {ref.start_line}"

    return False, ""


def lint_line_range(
    record: Record,
    record_idx: int,
) -> list[Diagnostic]:
    """Check if line/character ranges are valid."""
    diagnostics: list[Diagnostic] = []

    if record.file and record.file.start_line is not None:
        is_invalid, error_msg = _is_position_invalid(record.file)
        if is_invalid:
            diagnostics.append(Diagnostic(
                file=record._source_file,
                line=record._start_line,
                column=None,
                severity=Severity.ERROR,
                code=ErrorCode.E011,
                message=f"Invalid range in @file: {error_msg}",
                record_index=record_idx,
            ))

    if record.input and record.input.start_line is not None:
        is_invalid, error_msg = _is_position_invalid(record.input)
        if is_invalid:
            diagnostics.append(Diagnostic(
                file=record._source_file,
                line=record._start_line,
                column=None,
                severity=Severity.ERROR,
                code=ErrorCode.E011,
                message=f"Invalid range in @input: {error_msg}",
                record_index=record_idx,
            ))

    return diagnostics


def lint_reply_to(
    records: list[Record],
    source_file: Optional[Path],
) -> list[Diagnostic]:
    """Check @reply-to targets: must point at an @id in the same file; no cycles."""
    diagnostics: list[Diagnostic] = []

    id_to_idx: dict[str, int] = {}
    for idx, record in enumerate(records):
        if record.id and record.id not in id_to_idx:
            id_to_idx[record.id] = idx

    for idx, record in enumerate(records):
        if not record.reply_to:
            continue

        if record.reply_to not in id_to_idx:
            diagnostics.append(Diagnostic(
                file=source_file,
                line=record._start_line,
                column=None,
                severity=Severity.WARNING,
                code=WarningCode.W011,
                message=f"@reply-to points at unknown id: {record.reply_to}",
                record_index=idx,
            ))
            continue

        # Walk up the chain to detect cycles.
        seen = {idx}
        cursor = id_to_idx[record.reply_to]
        while True:
            if cursor in seen:
                diagnostics.append(Diagnostic(
                    file=source_file,
                    line=record._start_line,
                    column=None,
                    severity=Severity.WARNING,
                    code=WarningCode.W011,
                    message=f"@reply-to forms a cycle through: {record.reply_to}",
                    record_index=idx,
                ))
                break
            seen.add(cursor)
            parent = records[cursor].reply_to
            if not parent or parent not in id_to_idx:
                break
            cursor = id_to_idx[parent]

    return diagnostics


def lint_canonical_format(
    records: list[Record],
    original_text: str,
    file: Optional[Path],
) -> list[Diagnostic]:
    """Check if file is in canonical format."""
    diagnostics: list[Diagnostic] = []

    canonical = write_string(records, version_header=False)
    original_normalized = original_text.replace('\r\n', '\n')

    if original_normalized != canonical:
        diagnostics.append(Diagnostic(
            file=file,
            line=1,
            column=None,
            severity=Severity.WARNING,
            code=WarningCode.W008,
            message="Non-canonical formatting detected",
        ))

    return diagnostics


def lint_string(
    text: str,
    source_file: Optional[Path] = None,
    check_sources: bool = True,
    check_canonical: bool = True,
) -> ParseResult:
    """Lint a MarkBack string."""
    result = parse_string(text, source_file=source_file)

    for idx, record in enumerate(result.records):
        result.diagnostics.extend(lint_feedback_json(
            record.feedback,
            source_file,
            record._end_line,
            idx,
        ))

        if not record.feedback.startswith("json:"):
            result.diagnostics.extend(lint_feedback_structured(
                record.feedback,
                source_file,
                record._end_line,
                idx,
            ))

        if check_sources:
            base_path = source_file.parent if source_file else None
            result.diagnostics.extend(lint_file_exists(record, base_path, idx))
            result.diagnostics.extend(lint_input_exists(record, base_path, idx))

        result.diagnostics.extend(lint_line_range(record, idx))

    result.diagnostics.extend(lint_reply_to(result.records, source_file))

    if check_canonical and result.records and not result.has_errors:
        result.diagnostics.extend(lint_canonical_format(
            result.records,
            text,
            source_file,
        ))

    return result


def lint_file(
    path: Path,
    check_sources: bool = True,
    check_canonical: bool = True,
) -> ParseResult:
    """Lint a MarkBack file."""
    path = Path(path)
    try:
        text = path.read_text(encoding="utf-8")
    except UnicodeDecodeError:
        return ParseResult(
            records=[],
            diagnostics=[
                Diagnostic(
                    file=path,
                    line=None,
                    column=None,
                    severity=Severity.ERROR,
                    code=ErrorCode.E006,
                    message="File is not valid UTF-8",
                )
            ],
            source_file=path,
        )
    except FileNotFoundError:
        return ParseResult(
            records=[],
            diagnostics=[
                Diagnostic(
                    file=path,
                    line=None,
                    column=None,
                    severity=Severity.ERROR,
                    code=ErrorCode.E006,
                    message="File not found",
                )
            ],
            source_file=path,
        )

    return lint_string(
        text,
        source_file=path,
        check_sources=check_sources,
        check_canonical=check_canonical,
    )


def lint_files(
    paths: list[Path],
    check_sources: bool = True,
    check_canonical: bool = True,
) -> list[ParseResult]:
    """Lint multiple MarkBack files."""
    results: list[ParseResult] = []

    for path in paths:
        path = Path(path)
        if path.is_dir():
            for mb_file in path.glob("**/*.mb"):
                results.append(lint_file(
                    mb_file,
                    check_sources=check_sources,
                    check_canonical=check_canonical,
                ))
        else:
            results.append(lint_file(
                path,
                check_sources=check_sources,
                check_canonical=check_canonical,
            ))

    return results


def format_diagnostics(
    diagnostics: list[Diagnostic],
    format: str = "human",
) -> str:
    """Format diagnostics for output."""
    if format == "json":
        return json.dumps([d.to_dict() for d in diagnostics], indent=2)

    return '\n'.join(str(d) for d in diagnostics)


def summarize_results(results: list[ParseResult]) -> dict:
    """Summarize lint results."""
    total_records = sum(len(r.records) for r in results)
    total_errors = sum(r.error_count for r in results)
    total_warnings = sum(r.warning_count for r in results)
    files_with_errors = sum(1 for r in results if r.has_errors)
    files_with_warnings = sum(1 for r in results if r.has_warnings)

    return {
        "files": len(results),
        "records": total_records,
        "errors": total_errors,
        "warnings": total_warnings,
        "files_with_errors": files_with_errors,
        "files_with_warnings": files_with_warnings,
    }

File: packages/markbackjs/src/linter.ts
import fs from "fs";
import path from "path";
import { Diagnostic, ErrorCode, ParseResult, Record as MarkbackRecord, Severity, WarningCode } from "./types";
import { parseString } from "./parser";
import { writeRecordCanonical, writeRecordsMulti } from "./writer";

export interface LintOptions {
  sourceFile?: string | null;
  checkSources?: boolean;
  checkCanonical?: boolean;
}

function lintFeedbackJson(
  feedback: string,
  file?: string | null,
  line?: number | null,
  recordIdx?: number | null,
): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];

  if (feedback.startsWith("json:")) {
    const jsonStr = feedback.slice(5);
    try {
      JSON.parse(jsonStr);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      diagnostics.push(
        new Diagnostic({
          file: file ?? null,
          line: line ?? null,
          column: null,
          severity: Severity.ERROR,
          code: ErrorCode.E007,
          message: `Invalid JSON after json: prefix: ${message}`,
          recordIndex: recordIdx ?? null,
        }),
      );
    }
  }

  return diagnostics;
}

function lintFeedbackStructured(
  feedback: string,
  file?: string | null,
  line?: number | null,
  recordIdx?: number | null,
): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];
  let inQuote = false;
  let escaped = false;

  for (const char of feedback) {
    if (escaped) {
      escaped = false;
      continue;
    }
    if (char === "\\") {
      escaped = true;
      continue;
    }
    if (char === '"') {
      inQuote = !inQuote;
    }
  }

  if (inQuote) {
    diagnostics.push(
      new Diagnostic({
        file: file ?? null,
        line: line ?? null,
        column: null,
        severity: Severity.ERROR,
        code: ErrorCode.E008,
        message: "Unclosed quote in structured attribute value",
        recordIndex: recordIdx ?? null,
      }),
    );
  }

  return diagnostics;
}

function lintFileExists(record: MarkbackRecord, basePath: string | null, recordIdx: number): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];

  if (record.file && !record.file.isUri) {
    try {
      const resolved = record.file.resolve(basePath);
      if (!fs.existsSync(resolved)) {
        diagnostics.push(
          new Diagnostic({
            file: record._sourceFile ?? null,
            line: record._startLine ?? null,
            column: null,
            severity: Severity.WARNING,
            code: WarningCode.W003,
            message: `@file not found: ${record.file}`,
            recordIndex: recordIdx,
          }),
        );
      }
    } catch (_err) {
      // Ignore URIs that cannot be resolved to paths.
    }
  }

  return diagnostics;
}

function lintInputExists(record: MarkbackRecord, basePath: string | null, recordIdx: number): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];

  if (record.input && !record.input.isUri) {
    try {
      const resolved = record.input.resolve(basePath);
      if (!fs.existsSync(resolved)) {
        diagnostics.push(
          new Diagnostic({
            file: record._sourceFile ?? null,
            line: record._startLine ?? null,
            column: null,
            severity: Severity.WARNING,
            code: WarningCode.W009,
            message: `@input not found: ${record.input}`,
            recordIndex: recordIdx,
          }),
        );
      }
    } catch (_err) {
      // Ignore URIs that cannot be resolved to paths.
    }
  }

  return diagnostics;
}

interface PositionCheck {
  isInvalid: boolean;
  errorMsg: string;
}

function isPositionInvalid(ref: { startLine: number | null; endLine: number | null; startColumn: number | null; endColumn: number | null }): PositionCheck {
  if (ref.startLine === null || ref.endLine === null) {
    return { isInvalid: false, errorMsg: "" };
  }

  if (ref.endLine < ref.startLine) {
    return { isInvalid: true, errorMsg: `end line ${ref.endLine} is less than start line ${ref.startLine}` };
  }

  if (ref.endLine === ref.startLine) {
    if (ref.startColumn !== null && ref.endColumn !== null && ref.endColumn < ref.startColumn) {
      return { isInvalid: true, errorMsg: `end column ${ref.endColumn} is less than start column ${ref.startColumn} on line ${ref.startLine}` };
    }
  }

  return { isInvalid: false, errorMsg: "" };
}

function lintLineRange(record: MarkbackRecord, recordIdx: number): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];

  if (record.file && record.file.startLine !== null) {
    const { isInvalid, errorMsg } = isPositionInvalid(record.file);
    if (isInvalid) {
      diagnostics.push(
        new Diagnostic({
          file: record._sourceFile ?? null,
          line: record._startLine ?? null,
          column: null,
          severity: Severity.ERROR,
          code: ErrorCode.E011,
          message: `Invalid range in @file: ${errorMsg}`,
          recordIndex: recordIdx,
        }),
      );
    }
  }

  if (record.input && record.input.startLine !== null) {
    const { isInvalid, errorMsg } = isPositionInvalid(record.input);
    if (isInvalid) {
      diagnostics.push(
        new Diagnostic({
          file: record._sourceFile ?? null,
          line: record._startLine ?? null,
          column: null,
          severity: Severity.ERROR,
          code: ErrorCode.E011,
          message: `Invalid range in @input: ${errorMsg}`,
          recordIndex: recordIdx,
        }),
      );
    }
  }

  return diagnostics;
}

function lintReplyTo(records: MarkbackRecord[], sourceFile?: string | null): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];

  const idToIdx: { [key: string]: number } = {};
  records.forEach((record, idx) => {
    if (record.id && idToIdx[record.id] === undefined) {
      idToIdx[record.id] = idx;
    }
  });

  records.forEach((record, idx) => {
    if (!record.replyTo) {
      return;
    }

    if (idToIdx[record.replyTo] === undefined) {
      diagnostics.push(
        new Diagnostic({
          file: sourceFile ?? null,
          line: record._startLine ?? null,
          column: null,
          severity: Severity.WARNING,
          code: WarningCode.W011,
          message: `@reply-to points at unknown id: ${record.replyTo}`,
          recordIndex: idx,
        }),
      );
      return;
    }

    // Walk up the chain to detect cycles.
    const seen = new Set<number>([idx]);
    let cursor = idToIdx[record.replyTo];
    while (true) {
      if (seen.has(cursor)) {
        diagnostics.push(
          new Diagnostic({
            file: sourceFile ?? null,
            line: record._startLine ?? null,
            column: null,
            severity: Severity.WARNING,
            code: WarningCode.W011,
            message: `@reply-to forms a cycle through: ${record.replyTo}`,
            recordIndex: idx,
          }),
        );
        break;
      }
      seen.add(cursor);
      const parent = records[cursor].replyTo;
      if (!parent || idToIdx[parent] === undefined) {
        break;
      }
      cursor = idToIdx[parent];
    }
  });

  return diagnostics;
}

function lintCanonicalFormat(records: MarkbackRecord[], originalText: string, file?: string | null): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];

  const canonical =
    records.length === 1 ? `${writeRecordCanonical(records[0])}\n` : writeRecordsMulti(records);
  const originalNormalized = originalText.replace(/\r\n/g, "\n").replace(/\r/g, "\n");

  if (originalNormalized !== canonical) {
    diagnostics.push(
      new Diagnostic({
        file: file ?? null,
        line: 1,
        column: null,
        severity: Severity.WARNING,
        code: WarningCode.W008,
        message: "Non-canonical formatting detected",
      }),
    );
  }

  return diagnostics;
}

class InvalidUtf8Error extends Error {
  code: string;

  constructor(message: string) {
    super(message);
    this.code = "ERR_INVALID_UTF8";
  }
}

function readUtf8FileSync(filePath: string): string {
  const data = fs.readFileSync(filePath);
  const decoder = new TextDecoder("utf-8", { fatal: true });
  try {
    return decoder.decode(data);
  } catch (_err) {
    throw new InvalidUtf8Error("File is not valid UTF-8");
  }
}

export function lintString(text: string, options: LintOptions = {}): ParseResult {
  const sourceFile = options.sourceFile ?? null;
  const checkSources = options.checkSources ?? true;
  const checkCanonical = options.checkCanonical ?? true;

  const result = parseString(text, sourceFile);

  result.records.forEach((record, idx) => {
    result.diagnostics.push(...lintFeedbackJson(record.feedback, sourceFile, record._endLine, idx));

    if (!record.feedback.startsWith("json:")) {
      result.diagnostics.push(...lintFeedbackStructured(record.feedback, sourceFile, record._endLine, idx));
    }

    if (checkSources) {
      const basePath = sourceFile ? path.dirname(sourceFile) : null;
      result.diagnostics.push(...lintFileExists(record, basePath, idx));
      result.diagnostics.push(...lintInputExists(record, basePath, idx));
    }

    result.diagnostics.push(...lintLineRange(record, idx));
  });

  result.diagnostics.push(...lintReplyTo(result.records, sourceFile));

  if (checkCanonical && result.records.length > 0 && !result.hasErrors) {
    result.diagnostics.push(...lintCanonicalFormat(result.records, text, sourceFile));
  }

  return result;
}

export function lintFile(filePath: string, options: Omit<LintOptions, "sourceFile"> = {}): ParseResult {
  try {
    const text = readUtf8FileSync(filePath);
    return lintString(text, { ...options, sourceFile: filePath });
  } catch (err) {
    if (err && typeof err === "object" && "code" in err && err.code === "ENOENT") {
      return new ParseResult([], [new Diagnostic({ file: filePath, line: null, column: null, severity: Severity.ERROR, code: ErrorCode.E006, message: "File not found" })], filePath);
    }
    if (err && typeof err === "object" && "code" in err && err.code === "ERR_INVALID_UTF8") {
      return new ParseResult([], [new Diagnostic({ file: filePath, line: null, column: null, severity: Severity.ERROR, code: ErrorCode.E006, message: "File is not valid UTF-8" })], filePath);
    }
    throw err;
  }
}

function walkFiles(dir: string): string[] {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  const files: string[] = [];

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...walkFiles(fullPath));
    } else if (entry.isFile()) {
      files.push(fullPath);
    }
  }

  return files;
}

export function lintFiles(paths: string[], options: Omit<LintOptions, "sourceFile"> = {}): ParseResult[] {
  const results: ParseResult[] = [];

  for (const inputPath of paths) {
    let stats: fs.Stats | null = null;
    try {
      stats = fs.statSync(inputPath);
    } catch (_err) {
      results.push(lintFile(inputPath, options));
      continue;
    }

    if (stats.isDirectory()) {
      const files = walkFiles(inputPath).sort();
      for (const file of files) {
        if (file.endsWith(".mb")) {
          results.push(lintFile(file, options));
        }
      }
    } else {
      results.push(lintFile(inputPath, options));
    }
  }

  return results;
}

export function formatDiagnostics(diagnostics: Diagnostic[], format: "human" | "json" = "human"): string {
  if (format === "json") {
    return JSON.stringify(diagnostics.map((d) => d.toDict()), null, 2);
  }

  return diagnostics.map((d) => d.toString()).join("\n");
}

export function summarizeResults(results: ParseResult[]): { [key: string]: number } {
  const totalRecords = results.reduce((sum, result) => sum + result.records.length, 0);
  const totalErrors = results.reduce((sum, result) => sum + result.errorCount, 0);
  const totalWarnings = results.reduce((sum, result) => sum + result.warningCount, 0);
  const filesWithErrors = results.filter((result) => result.hasErrors).length;
  const filesWithWarnings = results.filter((result) => result.hasWarnings).length;

  return {
    files: results.length,
    records: totalRecords,
    errors: totalErrors,
    warnings: totalWarnings,
    files_with_errors: filesWithErrors,
    files_with_warnings: filesWithWarnings,
  };
}
